runtime/compiler/optimizer/abstractinterpreter/AbsInterpreter.cpp

```diff
@@ -0,0 +1,3072 @@
+
+      for (auto e = block->getPredecessors().begin(); e != block->getPredecessors().end(); e++)
+         {
+         TR::Block* predBlock = (*e)->getFrom()->asBlock();
+
+         if (getState(predBlock) == NULL) //any block is not interpreted
+            {
+            allPresecessorsInterpreted = false;
+            if (oneInterpretedBlock)
+               break;
+            }
+         else  //any block is interpreted
+            {
+            oneInterpretedBlock = predBlock;
+            }
+         }
+
+      /*** Case 3.1: All predecessors are interpreted. Merge their states ***/
+      if (allPresecessorsInterpreted)
+         {
+         TR::Block* firstPredBlock = (*block->getPredecessors().begin())->getFrom()->asBlock();
+         TR::AbsStackMachineState* state = getState(firstPredBlock)->clone(region());
+
+         for (auto e = ++block->getPredecessors().begin(); e != block->getPredecessors().end(); e ++)
+            {
+            TR::Block* predBlock = (*e)->getFrom()->asBlock();
+            state->merge(getState(predBlock), region());
+            }
+
+         setState(block, state);
+         return;
+         }
+
+      /*** Case 3.2: some blocks not interpreted ***/
+      if (insideLoop && !allPresecessorsInterpreted && oneInterpretedBlock)
+         {
+         TR::AbsStackMachineState* state = NULL;
+
+         for (auto e = ++block->getPredecessors().begin(); e != block->getPredecessors().end(); e ++)
+            {
+            TR::Block* predBlock = (*e)->getFrom()->asBlock();
+
+            if (!state && getState(predBlock))
+               state = getState(predBlock)->clone(region());
+            else if (getState(predBlock))
+               state->merge(getState(predBlock), region());
+            }
+
+         setState(block, state);
+         return;
+         }
+
+      /*** Case 3.3: set the state to TOP ***/
+      if (!allPresecessorsInterpreted && oneInterpretedBlock)
+         {
+         TR::AbsStackMachineState* copiedState = getState(oneInterpretedBlock)->clone(region());
+         for (size_t i = 0; i < copiedState->getArraySize(); i ++)
+            {
+            copiedState->at(i) ? copiedState->at(i)->setToTop() : copiedState->set(i, new (region()) TR::AbsVPValue(vp(), NULL, TR::NoType));
+            }
+
+         TR::vector<TR::AbsValue*, TR::Region&> temp(region());
+         while (copiedState->getStackSize() != 0)
+            {
+            TR::AbsValue* val = copiedState->pop();
+            if (val)
+               val->setToTop();
+            else
+               val = new (region()) TR::AbsVPValue(vp(), NULL, TR::NoType);
+
+            temp.push_back(val);
+            }
+
+         for (size_t i = 0; i < temp.size(); i ++)
+            {
+            copiedState->push(temp[i]);
+            }
+
+         setState(block, copiedState);
+         return;
+         }
+
+      /*** Case 3.4: None of the predecessors is interpreted. Block is in dead code area. Do nothing ***/
```

Comment:
I think the reverse postorder traversal doesn't work for this special case is because there is a cycle in the CFG. Actually, this situation should be handled by our next PR which is handling loop, however, we decided to merge those code into this PR.

To handle the loop, we will determine if the code to be interpreted contains any back edges. If yes, we will call `TR_RegionAnalysis::getRegions()` to do the region analysis. This modified region analysis will accept our existing CFG and calculate the dominators by our modified `TR_Dominators::TR_Dominators()`. Then, for each region from the entry node, it will interpret them in the order that the region analysis gave. If it encounters any node which one's structure is block, it will run `TR::AbsInterpreter::interpretBlock()`. Otherwise, if the the node's structure is region, it will create a `RegionIterator` and traverse the blocks/regions in that regions in a reverse postorder,

Since we traverse all blocks and regions by following the given result of the region analysis, and the region analysis result are generated based on dominator, I don't think we will encounter the situation that we have to visit block E after an unreachable block B and before block D.